### PR TITLE
Add log for canonicalized plans in HBO

### PR DESCRIPTION
**Pull request by feilong-liu**
## Description
Log the stats equivalent query plan and the canonicalized plan from HBO.

## Motivation and Context
This is to help debug plan hash of HBO.

In HBO, a query will get a snapshot of the plan which we call stats equivalent plan. This plan is then canonicalized to get a new canonicalized plan, which later gets hashed to get a string. This hash string is used as map key to store and retrieve history.

However, one problem observed is that, for the same query, it can have different hash string. This is hard to debug, as it cannot be deterministically reproduced. In this PR, I log the stats equivalent and canonicalized plan, so as to help compare and debug such cases.

## Impact
Will help debug HBO hash generation.

## Test Plan
Tested locally end to end.

## Contributor checklist

- [ ] Please make sure your submission complies with our [development](https://github.com/prestodb/presto/wiki/Presto-Development-Guidelines#development), [formatting](https://github.com/prestodb/presto/wiki/Presto-Development-Guidelines#formatting), [commit message](https://github.com/prestodb/presto/wiki/Review-and-Commit-guidelines#commit-formatting-and-pull-requests), and [attribution guidelines](https://github.com/prestodb/presto/wiki/Review-and-Commit-guidelines#attribution).
- [ ] PR description addresses the issue accurately and concisely.  If the change is non-trivial, a GitHub Issue is referenced.
- [ ] Documented new properties (with its default value), SQL syntax, functions, or other functionality.
- [ ] If release notes are required, they follow the [release notes guidelines](https://github.com/prestodb/presto/wiki/Release-Notes-Guidelines).
- [ ] Adequate tests were added if applicable.
- [ ] CI passed.

## Release Notes
Please follow [release notes guidelines](https://github.com/prestodb/presto/wiki/Release-Notes-Guidelines) and fill in the release notes below.

```
== RELEASE NOTES ==

General Changes
* Add log of stats equivalent plan and canonicalized plan for HBO. This feature is controlled by session property `log_query_plans_used_in_history_based_optimizer`
```

### CODE REVIEW

File: presto-main/src/main/java/com/facebook/presto/cost/HistoryBasedPlanStatisticsCalculator.java [hunk 1]
```diff
@@ -150,17 +153,23 @@ public Supplier<HistoryBasedPlanStatisticsProvider> getHistoryBasedPlanStatistic
...
         if (!useHistoryBasedPlanStatisticsEnabled(session) || !plan.getStatsEquivalentPlanNode().isPresent()) {
             return ImmutableMap.of();
         }
 
         PlanNode statsEquivalentPlanNode = plan.getStatsEquivalentPlanNode().get();
         ImmutableMap.Builder<PlanCanonicalizationStrategy, PlanNodeWithHash> allHashesBuilder = ImmutableMap.builder();
+        if (recordPlan) {
+            Optional<String> statsEquivalentPlanString = serializeStatsEquivalentPlan(statsEquivalentPlanNode);
+            if (statsEquivalentPlanString.isPresent()) {
+                historyBasedStatisticsCacheManager.setStatsEquivalentPlan(session.getQueryId(), statsEquivalentPlanString.get());
```
Comment (feilong-liu) [inline presto-main/src/main/java/com/facebook/presto/cost/HistoryBasedPlanStatisticsCalculator.java:168]: Record stats equivalent plan

File: presto-main/src/main/java/com/facebook/presto/sql/analyzer/FeaturesConfig.java [hunk 2]
```diff
@@ -104,6 +104,7 @@ public class FeaturesConfig
     private int historyCanonicalPlanNodeLimit = 1000;
     private Duration historyBasedOptimizerTimeout = new Duration(10, SECONDS);
     private String historyBasedOptimizerPlanCanonicalizationStrategies = "IGNORE_SAFE_CONSTANTS";
+    private boolean logPlansUsedInHBO = true;
```
Comment (jaystarshot) [inline presto-main/src/main/java/com/facebook/presto/sql/analyzer/FeaturesConfig.java:107]: Since this going to add substantial weight to the stored entry, better to have this off by default.

**Comment by steveburnett** [comment]
"This feature is controlled by session property `log_query_plans_used_in_history_based_optimizer`. 

Suggest adding doc for the session property. Would http://prestodb.io/docs/current/plugin/redis-hbo-provider.html or http://prestodb.io/docs/current/admin/properties.html#optimizer-properties be the correct place for it?

**Review by jaystarshot — APPROVED** [review]

File: presto-docs/src/main/sphinx/admin/properties.rst [hunk 3]
```diff
@@ -811,6 +811,14 @@ Optimizer Properties
 
     Track history based plan statistics from complete plan fragments in failed queries.
 
+``optimizer.log-plans-used-in-hbo``
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+    * **Type:** ``boolean``
+    * **Default value:** ``false``
+
+    Log stats equivalent plan and canonicalized plans used in history based optimization.
```
Comment (steveburnett) [inline presto-docs/src/main/sphinx/admin/properties.rst:820]:
```suggestion
    Log the stats equivalent plan and canonicalized plans used in history based optimization.
```
"Log stats equivalent plan" seems a little hard to parse. Suggestion of "the" taken from the Description of the PR makes it easier to understand.

**Review by steveburnett — CHANGES_REQUESTED** [review]
Thanks for the doc! One tiny nit suggestion for readability.

File: presto-docs/src/main/sphinx/admin/properties.rst [hunk 4]
```diff
@@ -811,6 +811,14 @@ Optimizer Properties
 
     Track history based plan statistics from complete plan fragments in failed queries.
 
+``optimizer.log-plans-used-in-hbo``
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
```
Comment (steveburnett) [inline presto-docs/src/main/sphinx/admin/properties.rst:815]:
```suggestion
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
```
Suggest changing the heading formatting to the same character length as the character length in line 814.

File: presto-docs/src/main/sphinx/admin/properties.rst (hunk 3 above)
Comment (steveburnett) [inline presto-docs/src/main/sphinx/admin/properties.rst:820]:
```suggestion
    Log the stats equivalent plan and canonicalized plans used in history based optimization.
```
Let me know if this change doesn't work for you, I'd like to understand why.

**Review by steveburnett — CHANGES_REQUESTED** [review]
Thanks for the quick response! Two small nits.

**Review by steveburnett — APPROVED** [review]
LGTM! (docs) 

Pull updated branch, new local build,looks good. Thanks for the rapid response!

**Review by jaystarshot — APPROVED** [review]

File: presto-main/src/main/java/com/facebook/presto/cost/HistoryBasedPlanStatisticsCalculator.java [hunk 5]
```diff
@@ -37,6 +37,7 @@
 import static com.facebook.presto.SystemSessionProperties.getHistoryBasedOptimizerTimeoutLimit;
 import static com.facebook.presto.SystemSessionProperties.getHistoryInputTableStatisticsMatchingThreshold;
 import static com.facebook.presto.SystemSessionProperties.isVerboseRuntimeStatsEnabled;
+import static com.facebook.presto.SystemSessionProperties.logQueryPlansUsedInHBO;
```
[the remark this answers is not in the document]
Comment (feilong-liu) [reply presto-main/src/main/java/com/facebook/presto/cost/HistoryBasedPlanStatisticsCalculator.java]: Changed to logQueryPlansUsedInHistoryBasedOptimizer

File: presto-main/src/main/java/com/facebook/presto/sql/planner/CachingPlanCanonicalInfoProvider.java [hunk 6]
```diff
@@ -60,20 +60,20 @@ public CachingPlanCanonicalInfoProvider(HistoryBasedStatisticsCacheManager histo
     }
 
     @Override
-    public Optional<String> hash(Session session, PlanNode planNode, PlanCanonicalizationStrategy strategy, boolean cacheOnly)
+    public Optional<String> hash(Session session, PlanNode planNode, PlanCanonicalizationStrategy strategy, boolean cacheOnly, boolean recordPlan)
```
[the remark this answers is not in the document]
Comment (feilong-liu) [reply presto-main/src/main/java/com/facebook/presto/sql/planner/CachingPlanCanonicalInfoProvider.java]: Removed the record argument.

**Review by pranjalssh — APPROVED** [review]
LGTM
